Make minimal changes to CodeQL docs for new diagnostic data

felicitymay opened the pull request:
### Summary

This pull request makes some small updates to the CodeQL docs for the changes coming in CodeQL CLI 2.5.6. Please see the linked internal issue for details.

hmakholm commented on docs/codeql/codeql-cli/using-custom-queries-with-the-codeql-cli.rst:
```diff
@@ -33,8 +33,8 @@ following two properties to ensure that the results are interpreted correctly:
   digits, delimited by ``/`` or ``-``, identifying and classifying the query. 
-- Query type (``@kind``): identifies the query is an alert (``@kind problem``)
-  or a path (``@kind path-problem``).
+- Query type (``@kind``): identifies the query is an alert (``@kind problem``),
+  a path (``@kind path-problem``), or a diagnostic metric (``@kind diagnostic``).
```
Might as well take this opportunity to fix the not very helpful description of path-problem here:
```suggestion
- Query type (``@kind``): identifies the query is a simple alert (``@kind problem``),
  an alert documented by a sequence of code locations (``@kind path-problem``), or a diagnostic metric (``@kind diagnostic``).
```

adityasharad replied:
There's one more kind: a query with `@kind metric` with `@tags summary` is considered a "summary metric" query.
For `@kind diagnostic` I would use the term "extractor diagnostic" or simply "diagnostic", rather than "diagnostic metric".

hmakholm commented on docs/codeql/writing-codeql-queries/about-codeql-queries.rst:
```diff
@@ -58,6 +58,7 @@ Query metadata is used to identify your custom queries when they are added to th
     Queries that are contributed to the open source repository, added to a query pack in LGTM, or used to analyze a database with the :ref:`CodeQL CLI <codeql-cli>` must have a query type (``@kind``) specified. The ``@kind`` property indicates how to interpret and display the results of the query analysis:
 
     - Alert query metadata must contain ``@kind problem``.
+    - Diagnostic query metadata must contain ``@kind diagnostic``.
     - Path query metadata must contain ``@kind path-problem``.
```
Would be slightly nicer to put "diagnostic" either first or last so it doesn't some _between_ the alert kinds.

felicitymay replied:
Good point - I added it alphabetically, but am happy to change the order.

adityasharad commented on docs/codeql/codeql-cli/analyzing-databases-with-the-codeql-cli.rst:
```diff
@@ -142,6 +143,13 @@ These are stored alongside the code scanning suites with names of the form: ``<l
+Diagnostic information
+......................
+
+The code scanning query suites include additional diagnostic queries. When the database analysis is complete, the CLI generates the results file and reports any diagnostic data to standard output. If you choose to generate SARIF output, the diagnostic data is also included as `notification objects <https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.html#_Toc34317894>`__ in the SARI ...
```
There are two kinds of diagnostic data we introduced:
- summary metrics are stored in the properties bag on each `run` object, in a new property called `metricResults`. (This is our own addition to the SARIF spec. I don't think we have to document in detail exactly what the format is yet, in case we need to adjust it.)
- extractor diagnostics are `notification` objects in `toolExecutionNotifications`, as you describe.

felicitymay replied:
Hmm. Thank you for explaining. The information here was taken from the [changelog](https://github.com/github/semmle-code/blob/main/odasa-buildutils/codeql-resources/CHANGELOG.md#features-added).

Based on your comment above, I've generalized this section and removed the information about notification objects for now.

adityasharad commented on docs/codeql/writing-codeql-queries/about-codeql-queries.rst:
```diff
@@ -58,6 +58,7 @@ Query metadata is used to identify your custom queries when they are added to th
 
     - Alert query metadata must contain ``@kind problem``.
+    - Diagnostic query metadata must contain ``@kind diagnostic``.
     - Path query metadata must contain ``@kind path-problem``.
```
One more:
- Summary metric query metadata must contain `@kind metric` and `@tags summary`.

adityasharad commented on docs/codeql/writing-codeql-queries/about-codeql-queries.rst:
```diff
@@ -114,6 +115,8 @@ You can modify the alert message defined in the final column of the ``select`` s
 
 Select clauses for path queries (``@kind path-problem``) are crafted to display both an alert and the source and sink of an associated path graph. For more information, see ":doc:`Creating path queries <creating-path-queries>`."
 
+Select clauses for diagnostic queries (``@kind diagnostic``) have different requirements. For examples, see the `diagnostic queries in the CodeQL repository <https://github.com/github/codeql/search?q=%22%40kind+diagnostic%22>`__.
```
You may want to add a similar link for summary metric queries, though the search is a little harder to phrase because there are queries with `@kind metric` but no `@tags summary`, and we don't want to include those. https://github.com/github/codeql/blob/main/cpp/ql/src/Summary/LinesOfCode.ql is an example.

felicitymay commented on docs/codeql/writing-codeql-queries/about-codeql-queries.rst:
```diff
@@ -114,6 +116,8 @@ You can modify the alert message defined in the final column of the ``select`` s
 
 Select clauses for path queries (``@kind path-problem``) are crafted to display both an alert and the source and sink of an associated path graph. For more information, see ":doc:`Creating path queries <creating-path-queries>`."
 
+Select clauses for diagnostic queries (``@kind diagnostic``) and summary metric queries (``@kind metric`` and ``@tags summary``) have different requirements. For examples, see the `diagnostic queries <https://github.com/github/codeql/search?q=%22%40kind+diagnostic%22>`__ and the `summary metric queries <https://github.com/github/codeql/search?q=%22%40kind+metric%22+%22%40tags+summary%22>`__  in t ...
```
@adityasharad - I wrote this without knowing what the requirements were. You kindly pointed me to https://github.com/github/codeql-coreql-team/issues/816 and https://github.com/github/codeql-coreql-team/issues/817. Shall we update this section to be more explicit or are you happy to leave this with just the links to examples for now?

felicitymay commented:
@hmakholm and @adityasharad - many thanks for the timely reviews ✨ 

I've updated the PR to address your comments and regenerated the preview to reflect those changes.

hubwriter commented on docs/codeql/codeql-cli/analyzing-databases-with-the-codeql-cli.rst:
```diff
@@ -142,6 +143,13 @@ These are stored alongside the code scanning suites with names of the form: ``<l
+Diagnostic and summary information
+..................................
+
+The code scanning query suites include additional queries to report diagnostic data generated by the extractor during database creation and summary metrics for the CodeQL database generated. When the database analysis command completes, the CLI generates the results file and reports any diagnostic and summary data to standard output. If you choose to generate SARIF output, the additional data is  ...
```
I had to read the long first sentence twice. I think it would benefit from a comma after "during database creation".

felicitymay replied:
Sentence revised.

hubwriter reviewed (approved):
LGTM 👍 

Your changes don't quite match the documentation plan (https://github.com/github/docs-content/issues/4433#issuecomment-853893214) which mentions updating "Creating CodeQL databases" and adding a troubleshooting section. Did you change your mind about this?

felicitymay commented:
Thanks for the review @hubwriter 🙇🏻‍♀️ 

I'll revise that sentence 👍🏻 

> our changes don't quite match the documentation plan which mentions updating "Creating CodeQL databases" and adding a troubleshooting section

You're right. I should have called that out. I hadn't realized that the information on the success of the new CodeQL database is only available after you've run some analysis on the database. So the information belongs in the article on analyzing CodeQL databases instead. I guess we could add a link and mention the diagnostic data to the article on creating CodeQL databases 🤔 - that might be a good idea.

hubwriter commented:
Apologies for assigning myself this. I got mixed up on the Review Board about the Assignees column. I believe this is intended to show the author of the PR. I mistakenly thought it showed the reviewer, before noticing the Reviewer column. 😆

aeisenberg commented on docs/codeql/codeql-cli/analyzing-databases-with-the-codeql-cli.rst:
```diff
@@ -142,6 +143,13 @@ These are stored alongside the code scanning suites with names of the form: ``<l
+Diagnostic and summary information
+..................................
+
+When you create a CodeQL database, the extractor stores diagnostic data in the database. The code scanning query suites include additional queries to report on this diagnostic data and calculate summary metrics. When the database analysis command completes, the CLI generates the results file and reports any diagnostic and summary data to standard output. If you choose to generate SARIF output, th ...
```
```suggestion
When you create a CodeQL database, the extractor stores diagnostic data in the database. The code scanning query suites include additional queries to report on this diagnostic data and calculate summary metrics. When the database analysis completes, the CLI generates the results file and reports any diagnostic and summary data to standard output. If you choose to generate SARIF output, the additional data is also included in the SARIF file.
```

or 

```suggestion
When you create a CodeQL database, the extractor stores diagnostic data in the database. The code scanning query suites include additional queries to report on this diagnostic data and calculate summary metrics. When the ``database analyze`` command completes, the CLI generates the results file and reports any diagnostic and summary data to standard output. If you choose to generate SARIF output, the additional data is also included in the SARIF file.
```

felicitymay replied:
Thanks for the suggestion ✨

aeisenberg reviewed (commented):
This looks great. One tiny suggestion.

felicitymay commented:
Many thanks to everyone for their reviews - much appreciated.

I _think_ that I've addressed all feedback and that this is now ready for merging so that the changes are ready to publish as part of the CodeQL microsite update for CodeQL CLI 2.5.6.

felicitymay commented:
Since this repository dismisses stale reviews when you add new commits, I'll need an approving review 😄

felicitymay commented:
Thank you @aeisenberg ✨
